File: src/main/ipc/handlers/mod-bisect.ts
import type { NahidaDesktop } from "@main/index";
import { rh } from "@main/ipc/helper";
import { toErrorMessage } from "@shared/utils";

export function registerModBisectHandlers(d: NahidaDesktop) {
    rh("tools:bisectStart", async (game: string, excludePaths: string[] = []) => {
        try {
            return await d.service.modTools.modBisect.start(game, excludePaths);
        } catch (error) {
            d.logger.error(error, "tools:bisectStart");
            d.logger.error(
                {
                    channel: "tools:bisectStart",
                    stage: "start",
                    game,
                    excludePaths,
                    error: toErrorMessage(error),
                },
                "tools:bisectStart:context",
            );
            throw error;
        }
    });
    rh("tools:bisectValidateExcludePath", async (game: string, inputPath: string) => {
        try {
            return await d.service.modTools.modBisect.validateExcludePath(game, inputPath);
        } catch (error) {
            d.logger.error(error, "tools:bisectValidateExcludePath");
            d.logger.error(
                {
                    channel: "tools:bisectValidateExcludePath",
                    stage: "validate-exclude-path",
                    game,
                    inputPath,
                    error: toErrorMessage(error),
                },
                "tools:bisectValidateExcludePath:context",
            );
            throw error;
        }
    });
    rh("tools:bisectRespond", (fixed: boolean) => d.service.modTools.modBisect.respond(fixed));
    rh("tools:bisectUndoLastRound", () => d.service.modTools.modBisect.undoLastRound());
    rh("tools:bisectFinalize", (keepDisabled: string[]) =>
        d.service.modTools.modBisect.finalize(keepDisabled),
    );
    rh("tools:bisectCancel", () => d.service.modTools.modBisect.cancel());
    rh("tools:bisectGetState", () => d.service.modTools.modBisect.getState());
    rh("tools:bisectRecover", (game: string) => d.service.modTools.modBisect.recoverOrphans(game));
}

File: internal/tools/mod_bisect.go
package tools

import (
	"context"

	modbisect "nahida.live/desktop/internal/tools/mod_bisect"
)

type (
	BisectStatus   = modbisect.BisectStatus
	BisectSnapshot = modbisect.BisectSnapshot
)

func (t *Tools) BisectGetState() *BisectSnapshot {
	return t.bisect.BisectGetState()
}

func (t *Tools) BisectValidateExcludePath(ctx context.Context, game, inputPath string) (string, error) {
	return t.bisect.BisectValidateExcludePath(ctx, game, inputPath)
}

func (t *Tools) BisectStart(ctx context.Context, game string, excludePaths []string) (BisectSnapshot, error) {
	return t.bisect.BisectStart(ctx, game, excludePaths)
}

func (t *Tools) BisectRespond(ctx context.Context, fixed bool) (BisectSnapshot, error) {
	return t.bisect.BisectRespond(ctx, fixed)
}

func (t *Tools) BisectUndoLastRound(ctx context.Context) (BisectSnapshot, error) {
	return t.bisect.BisectUndoLastRound(ctx)
}

func (t *Tools) BisectFinalize(ctx context.Context, keepDisabled []string) (BisectSnapshot, error) {
	return t.bisect.BisectFinalize(ctx, keepDisabled)
}

func (t *Tools) BisectCancel(ctx context.Context) (BisectSnapshot, error) {
	return t.bisect.BisectCancel(ctx)
}

func (t *Tools) BisectRecover(ctx context.Context, game string) (int, error) {
	return t.bisect.BisectRecover(ctx, game)
}

// RecoverBisects restores interrupted sessions for every configured non-NTE game.
//
//wails:ignore
func (t *Tools) RecoverBisects(ctx context.Context) error {
	return t.bisect.RecoverBisects(ctx)
}

func (t *Tools) shutdownBisect() error {
	if t == nil || t.bisect == nil {
		return nil
	}
	return t.bisect.Shutdown()
}
